// Copyright 2019 Dell EMC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// The NP4IntelFPGA protobuf contains Netcope NP4 Intel FPGA configuration
// information passed down via the PI Node to the NP4 PI implementation.

syntax = "proto3";

option cc_generic_services = false;

package pi.np4;

message NP4Config {
  message Daemon {
    string addr = 1;
    uint32 port = 2;
  }
  oneof device {
    string path = 1;
    Daemon daemon = 2;
  }
  message Atom {
    uint32 id = 1;
  }
  Atom atom = 3;
  bytes fpga_bitstream = 4;
}

message DPDKConfig {
  string device_name = 1;  // In Domain:Bus:Device.Func format 0000:b1:00.0
  uint32 rx_ring_size = 2;
  uint32 tx_ring_size = 3;
  uint32 num_rx_rings = 4;
  uint32 num_tx_rings = 5;
  uint32 num_mbufs = 6;
  uint32 mbuf_cache_size = 7;
  uint32 burst_size = 8;
}

message P4DeviceConfig {
  uint64 node_id = 1;
  NP4Config np4_config = 2;
  DPDKConfig dpdk_config = 3;
}

